### PR TITLE
feat: Add execTemplate template function

### LINKED ISSUE #1619: New template function for capturing the output of the `template` function
## Is your feature request related to a problem? Please describe.

I have:

```console
❯ cat ~/.dotfiles/home/.chezmoitemplates/get-latest-github-release
{{- $url :=  printf "https://github.com/%s/releases/latest" . -}}
{{- output "curl" "-fsSL" "-o" "/dev/null" "-w" "%{url_effective}" $url | trim | base -}}
```

And I would like to use it like so:

```console
❯ cat ~/.dotfiles/home/.chezmoiexternal.yaml
{{ $denoVersion := template "get-latest-github-release" "denoland/deno" -}}
".deno/bin/deno":
  type: file
  url: "https://github.com/denoland/deno/releases/download/{{ $denoVersion }}/deno-x86_64-unknown-linux-gnu.zip"
  executable: true
  filter:
    command: zcat
```

But it's not possible, because:

```console
$ chezmoi execute-template '{{ $denoVersion := template "get-latest-github-release" "denoland/deno" }}{{ $denoVersion }}'
chezmoi: template: arg1:1: unexpected <template> in command
```

EDIT 1: I previously posted with the wrong error message above, which turns to be valid but posterior to the error originally reported, as pointed by @twpayne at [here](https://github.com/twpayne/chezmoi/issues/1619#issuecomment-967649423).

EDIT 2: In the variable assignment, it was missing `:` before `=` as pointed by @twpayne at [here](https://github.com/twpayne/chezmoi/issues/1619#issuecomment-967688720). But that wasn't the cause of the issue, anyway.

## Describe the solution you'd like

A function which would let me do that, as pointed in https://stackoverflow.com/a/40170999/12156188. So that I could do:

```console
$ chezmoi execute-template '{{ $denoVersion := execTemplate "get-latest-github-release" "denoland/deno" }}{{ $denoVersion }}'
v1.16.1
```

## Describe alternatives you've considered

I could not find any.

## Additional context

My dotfiles is currently bloated with a lot of code duplication, and now I'm trying to clean it up by using some templates.

### PR DESCRIPTION
**Pull request by felipecrs**
Fixes #1619

### CODE REVIEW

**Comment by felipecrs** [comment]
@twpayne I'm sure this is breaking a million rules, but it's doing the trick. If you have any advice on how I can improve it, I would appreciate a lot.

File: internal/chezmoi/sourcestate.go [hunk 1]
```diff
@@ -636,9 +637,18 @@ func (s *SourceState) Encryption() Encryption {
 	return s.encryption
 }
 
+var tmpl *template.Template
+
+func execTemplateTemplateFunc(name string, data interface{}) (string, error) {
+	buf := &bytes.Buffer{}
+	err := tmpl.ExecuteTemplate(buf, name, data)
```
Comment (felipecrs) [inline internal/chezmoi/sourcestate.go:644]: The main problem is that `execTemplateTemplateFunc` needs to have access to the same `tmpl` object in use in the function below. That's the reason I could not put it in the `templatefuncs.go`, but maybe there is a way to do it better that I could not think of (I'm Go beginner :P).

**Comment by MuXiu1997** [comment]
It does break the rules, does it work this way here?
```go
func WithExecTemplateFunc(execTemplateFuncBuilder func(s *SourceState) template.FuncMap) SourceStateOption {
	return func(s *SourceState) {
		s.templateFuncs["execTemplate"] = execTemplateFuncBuilder(s)
	}
}
```

**Comment by felipecrs** [comment]
@MuXiu1997 I'm really sorry but I don't understand where I should put your snippet. :(

**Comment by MuXiu1997** [comment]
`WithExecTemplateFunc`  in `internal/chezmoi/sourcestate.go`, Then do not modify `WithTemplateFuncs`

And `execTemplateFuncBuilder` in `templatefuncs.go`

**Comment by felipecrs** [comment]
> `WithExecTemplateFunc` in `internal/chezmoi/sourcestate.go`, Then do not modify `WithTemplateFuncs`
> 
> And `execTemplateFuncBuilder` in `templatefuncs.go`

Thank you, will try it tomorrow.

**Comment by MuXiu1997** [comment]
I made some changes, if you don't like it, please let me know
#1661

**Comment by felipecrs** [comment]
> @felipecrs I made some changes, if you don't like it, please let me know
> #1661 

Of course I do like, thanks a lot.

**Comment by felipecrs** [comment]
Closed due to #1661.

**Pull request closed without merging**